File: ruby/book-store/book_store.rb
class BookStore
  DISCOUNTS = [0, 0, 0.05, 0.1, 0.2, 0.25]
  BASE_PRICE = 8

  def self.find_group(basket)
    count = basket.keys.size
    basket.keys.each { |i|
      basket[i] -= 1
      basket.delete(i) if basket[i] == 0
    }
    return count
  end

  def self.calculate_price(basket)
    basket = basket.each_with_object(Hash.new(0)) { |v, hash| hash[v] += 1 }
    groups = Hash.new(0)
    groups[find_group(basket)] += 1 until basket.empty?

    [groups[5], groups[3]].min.times {
      groups[5] -= 1
      groups[3] -= 1
      groups[4] += 2
    }
    groups.reduce(0) { |agg, (k, v)| agg + k * v * BASE_PRICE * (1 - DISCOUNTS[k]) }
  end
end

File: python/book-store/book_store.py
discounts = [1, .95, .90, .8, .75]
def total(basket):
    groups = []
    while len(basket) > 0:
        u = set(basket)
        groups.append(len(u))
        for b in u:
            basket.remove(b)

    while 3 in groups and 5 in groups:
        groups.remove(3)
        groups.remove(5)
        groups.append(4)
        groups.append(4)

    total = 0
    for size in groups:
        total += size * 800 * discounts[size - 1]
    return total
